### ISSUE TITLE
Deep Learning Compilers: A First Visit

### ISSUE THREAD

**Issue by connglli** [connglli/blog-notes#128]
> These notes are taken from the PDTS '21 paper "The Deep Learning Compiler: A Comprehensive Survey"

## Overview

DL compilers are compilers: they take high-level source code as input and output low-level machine code. 

``` txt
          +----------+
source -> | compiler | -> binary
          +----------+
```

However, DL compilers are domain-specific and distinct from traditional compilers in the following high-level aspects:
1. Input. The source code of a DL compiler is a model generated by DL frameworks like Tensorflow, PyTorch, and ONNX. The format of the model, for example the `.onnx` format, is not required to be a Turing-complete language like C and Java, but can be treated as a a domain-specific programming language (DSL). Such languages, either in the binary or text format, are often not human-readable, but definitely interpretable: they define the layers, parameters, and super-parameters of DL models. The task of DL compilers is to compile these formats into binary executables that can run model computations layer by layer.
2. Output: Even though the output is a binary executable (e.g., ELF in linux, Mach-O in Mac), the binary extensively relies on tensor instructions (e.g., Intel AVX-512), tensor cores (e.g., co-processors), or even DL-specific hardware like TPU for execution rather than scalar and vector instructions like traditional compilers.
3. Compilation and optimization. Since DL frameworks extensively relies on tensor manipulation (e.g., matrix multiplication) which is difficult to be expressed by traditional IRs, the compilation, optimizations, and IRs are especially designed around tensors. Specifically, DL compilers need tensor- and IO-intensive (or DL-oriented) IRs and optimization passes (e.g., layer optimization) to generate highly optimized, efficient binaries. DL compilers also embrace some traditional optimizations like DCE and CSE.

``` txt
          +----------+
model -> | compiler | -> DL-oriented binary
          +----------+
```

Well-known DL compilers include TVM, Tensor Comprehension, Glow.

> Before DL compilers, the generated model should be loaded by the associated DL framework and run by pure interpretation or by their dependent computation libraries like MKL and cuBLAS, specifically compiled using highly-optimized tensor and IO instructions. The performace are bounded by the runtime (e..g, Python) or aforementioned libraries.

> Besides using DL framework written by existing programming languages, is it possible to create a DL programming language (or DSL)? The language has native support for data loading, model building, etc., and the associated DL compiler can have more assumptions and aggressive optimizations. The pros are obvious, how about the cons? Perhaps it looks like Matlab?

## Why DL compilers?

With the development of deep learning, traditional general-purpose ISAs are extended to support chip-level tensor computation; DL-specific co-processors or even hardware are also created. Since traditional runtimes like Python or compilers like LLVM and GCC does not assume the input programs are tensor-oriented, they cannot fully leverage the tensor instructions to extreme or cannot support to run on DL-specific hardware.

## Architecture

DL compilers embrace the traditional compiler design consisting of frontends and backends; they also design multi-level IRs. All of these are different from traditional ones.

**Tensor**. Each tensor is represented by a block of memory (perhaps sparse arrays because a large number of parameters of a model are 0, see Lottery Ticket Hypothesis). Besides the memory for each tensor element, there is a placeholder (metadata is a better name?) describing the shape of the tensor (maybe unknown in early stage). The placeholders are extensively used in optimizations; they help seperate the computation definition from the exact exections. During optimizations, the unknown placeholders would be inferred.

**High-level IR - Graph IR**. Different DL compilers have different IRs just like traditional compilers. But in general, all DL compilers first compiles the model to a similar high-level, hardware-independent IR, namely graph IR. Different from traditional graph-based IRs like the Ideal Graph in HotSpot and Graal, graph IR in DL compilers represent computations, especially tensor computations. The graph represented by graph IR is often termed *computation graph*. Computations within computation graph are based on some atomic operators, e.g., add, times, conv, pool, etc. In the graph, each node represents an atomic operator and the incomming/outgoing edges represent input/output tensors.

**Operators**. Operators represeted by nodes can be classified into the following categories 
- Algebraic operators. Such as additive, substraction, times and dot-multiplication.
- Neural network operators. Such as conv, poolling.
- Tensor operators. Such as reshape.
- Broadcast and reduction operators. Such as max and argmin.
- Derivative operator. This operator is required if the DL compiler would like to support model training.
- Control-flow operators. Operators such as `if` or `while` are still needed in DL compilers because some models like RNN requires recurrent relations and data-dependent conditional executions. Otherwise, such models have to be flattened (unrolled) at compile-time to remove such dependants.

**Hardware-independant, graph IR optimizations**.
+ Node-level optimizations considering a single node.
  - Nop elimination. Remove nodes which is redudant like summing but only one tensor.
  - Zero-dim-tensor elimination. Operations on zero-dim tensors can be safely eliminated like argmin C where the shape of C is `(0, 5, 3)`.
+ Block-level optimizations considering a sequence of nodes, often relying on mathmatical rules like communitivity and associativety.
  - Constant folding like traditional compilers.
  - Strength reduction. E.g., replace expensive operator combinations by cheaper ones (A^T x B^T => (B x A)^T).
  - Operator fusion. TODO: learn more❓❓❓❓
  - Operator sinking. TODO: learn more❓❓❓❓
+ Graph-level optimizations consiering the whole graph.
  - CSE, DCE, and other traditional optimizations.
  - Static memory planning. Tensor manipulation is expensive in spaces; static memory planning aims to reuse as many memories as possible.
  - Layout transformation. Finding the best layout to store tensors. Some hardware requires specific data layout for an efficient computation.

**Low-level IRs**. There are some off-the-shelf low-level IRs and compilers which can be used by DL compilers.
- Halide IR and Halide.
- Polyhedral-based IRs.
- Other IRs.
TODO: Learn more about these IRs. ❓❓❓❓

**Hardware-dependent, low-level IR optimizations**. This enables efficient code generation.
+ Hardware intrinsic mapping. Map an DL operator to a hardware tensorization intrinsic.
+ Memory allocation and fetching. Find better memory allocation site to achieve faster memory fetching because hardware may have hierarchical momories with different capacity and access latency.
+ Loop-oriented optimizations. Loop unrolling, fusion, reordering; tiling; and sliding window. TODO Learn more. These are different from traditional loop opts, especially fusion, sliding window. They also takes the cache into consideration, for example, tiling can split a 10-iter loop into outer 2-iter and 5-iter inner loops, where the inner can better fit into hardware caches. ❓❓❓❓

**Auto-tuning**. TODO I think the model are generated with fixed parameters, so why do we need auto-tuning???? Is this used in traininging??? ❓❓❓❓

**Code generation based on low-level IR**. Many DL compilers sink to the LLVM infra for code generation (LLVM suppors to design custom ISAs for unseen hardware). However, directly transforming the low-level IR to LLVM bitcode often makes the generated binary inefficient. To avoid this, prevalent DL compilers adopt the following two strategies:
1. Apply hardware-specific loop transformation before tranforming to LLVM bitcode.
2. Provide additional information about the hardware for the optimization passes.

## References

1. ISAs are progressively extended to support more features: from scalar, to float, vector, and even tensor.
2. [The Deep Learning Compiler: A Comprehensive Survey](https://arxiv.org/pdf/2002.03794v4.pdf)
